#163426579 -  Implement User Signup

teepha opened the pull request:
#### What does this PR do?
Create User Sign up page

#### Description of Task to be completed?
* [x] write tests for the feature
* [x] create the Input component
* [x] create the Signup page

#### How should this be manually tested?
- clone the repo
- cd into the directory
- cd into the folder and run `npm install` to install dependencies
- run `npm test`
- run `npm start`

#### What are the relevant pivotal tracker stories?
[#163426579](https://www.pivotaltracker.com/story/show/163426579)

<img width="1440" alt="Signup-page" src="https://user-images.githubusercontent.com/44369235/54367510-c6fa0180-4672-11e9-8af2-7ec1bd03ee63.png">

Commit:
feat(auth): implement user signup

- create the user signup page
- render the page

[Finishes #163426579]

Commit:
feat(auth): implement user signup

- create the user signup page
- render the page

[Finishes #163426579]

benfluleck reviewed (changes requested):
Please implement feedback

benfluleck commented on __tests__/HomePage.test.js:
```diff
@@ -1,6 +1,6 @@
 import React from "react";
 import Homepage from "@/pages/Homepage";
```
Please change this alias, this is not readable. You can create an alias for pages

benfluleck commented on __tests__/Input.test.js:
```diff
@@ -0,0 +1,69 @@
+
+  it("should allow us to set type props", () => {
+    const { type } = wrapper.props().children.props.children[0].props;
+    expect(wrapper.props());
```
this line is not testing anything?

benfluleck commented on __tests__/Input.test.js:
```diff
@@ -0,0 +1,69 @@
+import React from "react";
+import Input from "@/components/common/Input";
```
I told you to redo this alias

benfluleck commented on __tests__/Input.test.js:
```diff
@@ -0,0 +1,69 @@
+  });
+
+  it("should allow us to set handleChange props", () => {
+    const { handleChange } = wrapper.props().children.props.children[0].props;
```
Use jest.fn() to test that handleChange is called

benfluleck replied:
To achieve this you will need to find one of the inputs with wrapper.find() and then trigger the change event. Allowing the handleChange to be called cc @teepha

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+    margin: auto;
+    width: 100%;
+    height: 404px;
+    border: 1px solid #fff;
```
put the #fff in your color variables

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+      font-size: 0.98rem;
+      a {
+        margin-left: 0.4rem;
+        color: $text;
```
Change this to text-colors

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+
+.sign-up-page__btn {
+  background-color: $background !important;
+  color: $text !important;
```
Why are these already being set to important??

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+
+  .ui .input,
+  .sign-up-page__btn {
+    height: 30px !important;
```
This is the first page and you are already creating important tags, this is not a good sign

benfluleck commented on src/index.scss:
```diff
@@ -2,8 +2,81 @@
+
+.ui.black.basic.button {
+    
+    background: #FFFFFF;
```
This is not indented properly

benfluleck commented on src/index.scss:
```diff
@@ -2,8 +2,81 @@
+.ui.fluid.left.icon.input {
+  width: 299.51px;
+  height: 37.54px;
+  background: #ffffff;
```
All these colours on this page need to go into your colors.scss

benfluleck commented on src/index.scss:
```diff
@@ -2,8 +2,81 @@
+  width: 100%;
+  height: 100%;
+  background: rgba(0, 0, 0, 0.63);
+  /* background: url(iso-republic-hands-using-laptops.jpg); */
```
Remove this comment

benfluleck commented on src/index.scss:
```diff
@@ -2,8 +2,81 @@
 * {
   padding: 0;
   margin: 0;
+  font-family: Helvetica;
```
create a font.scss for your fonts

benfluleck commented on jest.config.js:
```diff
@@ -6,10 +6,15 @@ module.exports = {
     pages: "<rootDir>/src/pages",
     src: "<rootDir>/src",
-    redux: "<rootDir>/src/redux",
+    "^<Components>(.*)$": "<rootDir>/src/components$1",
```
`<components>` and `<assets>`

benfluleck commented on webpack/webpack.common.js:
```diff
@@ -39,13 +31,28 @@ module.exports = {
     alias: {
       "<Pages>": path.resolve(__dirname, "../src/pages"),
-      "<Components>": path.resolve(__dirname, "../src/components")
+      "<Components>": path.resolve(__dirname, "../src/components"),
```
Please change the casing on all of these

benfluleck commented on jest.config.js:
```diff
@@ -6,10 +6,15 @@ module.exports = {
   moduleNameMapper: {
     "@/(.*)$": path.join(projectDir, "src/$1"),
-    components: "<rootDir>/components",
+    components: "<rootDir>/src/components",
```
Talk to Ndifreke and ask what he is doing here so you can refactor this properly

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+        margin-left: 0.4rem;
+        color: $text;
+        &:hover {
+          color: $background;
```
background-color??? maybe give it a name

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+        padding-left: 1rem;
+      }
+      span {
+        color: white;
```
Don't mix named colors and hex colors

benfluleck commented on src/assets/styles/SignUpPage.scss:
```diff
@@ -0,0 +1,178 @@
+      color: $black;
+
+      &:hover {
+        color: $background;
```
Rename this to a more descriptive color

benfluleck commented on src/components/common/Input.js:
```diff
@@ -0,0 +1,41 @@
+          type={type || "text"}
+          name={name}
+          className={className}
+          placeholder={placeholder || "Enter input"}
```
Use defaultProps for this case

benfluleck commented on src/components/common/Input.js:
```diff
@@ -0,0 +1,41 @@
+    <div className="field">
+      <div className="ui left icon input">
+        <input
+          type={type || "text"}
```
Use default props for this case

benfluleck commented on src/components/common/Input.js:
```diff
@@ -0,0 +1,41 @@
+          className={className}
+          placeholder={placeholder || "Enter input"}
+          value={value}
+          onChange={handleChange}
```
Make this required and default to an empty function in defaultProps like () => {}

obasajujoshua31 committed:
feat(authsignup):  user should be able to signup

- implement social media authentication
- implement local authentication and email confirmation

[Delivers #163426579]

Commit:
feat(auth): implement user signup

- write tests for the components, pages and action

[Delivers #163426579]

benfluleck commented on __tests__/actions/auth/signup.test.js:
```diff
@@ -0,0 +1,47 @@
+import moxios from "moxios";
+import signUpUser from "<actions>/auth/signup";
+import actionTypes from "<actions>/auth/types";
+import userData from "../../__mock__/userData";
```
Please add the mocks to the alias

benfluleck commented on __tests__/setup.config.js:
```diff
@@ -1,5 +1,15 @@
+const middleware = [thunk];
+const storeMock = configureMockStore(middleware);
+
+global.React = React;
```
I do not agree with this you can not just attack all of these on the global object. IT already has enough objects and functions to worry about

benfluleck commented on jest.config.js:
```diff
@@ -1,15 +1,28 @@
-    pages: "<rootDir>/src/pages",
-    src: "<rootDir>/src",
-    redux: "<rootDir>/src/redux",
+    "\\.(svg)$": "<rootDir>/__tests__/__mock__",
```
Do you use svgs or pngs ?

benfluleck commented on package.json:
```diff
@@ -58,9 +60,20 @@
   "dependencies": {
+    "axios": "^0.18.0",
     "dotenv": "^6.2.0",
+    "jwt-decode": "^2.2.0",
```
What is this doing here?

benfluleck commented on src/api/request.js:
```diff
@@ -0,0 +1,21 @@
+    const response = await axios({
+      headers: {
+        "Access-Control-Allow-Credentials": true,
+        "Access-Control-Allow-Origin": "*",
```
What is this file doing? I asked @obasajujoshua31  and he could not tell me

benfluleck commented on src/components/Routes.jsx:
```diff
@@ -0,0 +1,32 @@
+import HomePage from "<pages>/Homepage";
+import SignUpWrapper from "<components>/auth/SignUp";
```
This file is not a jsx file as it doesn't rely on React. It is more a javascript file with a routes object. So it should be `routes.js`

benfluleck commented on src/components/common/Input.jsx:
```diff
@@ -0,0 +1,56 @@
+Input.propTypes = {
+  type: PropTypes.string,
+  name: PropTypes.string,
+  value: PropTypes.string,
```
This should be required

teepha replied:
I thought you said I should not make it required, that  I should set defaultValues instead

benfluleck commented on src/constants/constants.js:
```diff
@@ -0,0 +1,3 @@
+export const API_URL = "http://localhost:3000/api/v1";
```
should this be pointing to your local branches or heroku ?

benfluleck commented on src/pages/SignUpPage.jsx:
```diff
@@ -0,0 +1,94 @@
+};
+
+SignUpPage.propTypes = {
+  firstName: PropTypes.string,
```
For Signup firstName , lastName and email should be required

benfluleck commented on __tests__/pages/ProfilePage.test.js:
```diff
@@ -0,0 +1,14 @@
+describe("ProfilePage", () => {
+  it("should contain h1 element", () => {
+    const profilePageShallow = shallow(<ProfilePage />);
+    expect(profilePageShallow.contains(<h1>welcome to Profile</h1>)).toBe(
```
Please change this line to `expect(profilePageShallow.contains(<h1>welcome to Profile</h1>)).toBeTruthy()`

benfluleck commented on __tests__/utils/validation.test.js:
```diff
@@ -0,0 +1,116 @@
+    const validationResult = validateCreateForm(state, {
+      target: { name: "unknown", value: "gorilla" }
+    });
+    expect(validationResult.unknown).toEqual(undefined);
```
use `.toBeFalsy`

benfluleck commented on src/components/auth/AuthWrapper.jsx:
```diff
@@ -0,0 +1,51 @@
+    render() {
+      return (
+        <AuthComponent
+          state={this.state}
```
Please do not treat state like props. Rename this  or destructure the arguments

benfluleck commented on src/components/auth/SignUp.jsx:
```diff
@@ -0,0 +1,57 @@
+  handleChange: PropTypes.func,
+  handleSubmit: PropTypes.func,
+  errors: PropTypes.object,
+  state: PropTypes.object,
```
Remove this from the props check

benfluleck commented on src/components/auth/VerifyAccount.jsx:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  render() {
+    return null;
```
What is this?

benfluleck replied:
This goes against React principles. This probably is better in the route as `renderProps`

benfluleck commented on __tests__/index.js:
```diff
@@ -1 +1,11 @@
-import "./HomePage.test";
+import "./pages/HomePage.test";
+import "./pages/ProfilePage.test";
+// import "./components/auth/AuthWrapper.test";
```
What is this doing here ?

benfluleck commented on __tests__/pages/ProfilePage.test.js:
```diff
@@ -0,0 +1,14 @@
+describe("ProfilePage", () => {
+  it("should contain h1 element", () => {
+    const profilePageShallow = shallow(<ProfilePage />);
+    expect(
```
Space between the declarations and the test please

benfluleck commented on __tests__/reducers/authReducer.test.js:
```diff
@@ -0,0 +1,40 @@
+
+  it("should handle VERIFY_USER", () => {
+    const payload = "Verification Successful!";
+    expect(
```
Space here please

benfluleck commented on __tests__/reducers/authReducer.test.js:
```diff
@@ -0,0 +1,40 @@
+  });
+
+  it("should handle SET_CURRENT_USER", () => {
+    expect(
```
Space here please

benfluleck commented on __tests__/reducers/authReducer.test.js:
```diff
@@ -0,0 +1,40 @@
+
+describe("AuthReducer test", () => {
+  it("should return the initial state ", () => {
+    expect(reducers(undefined, {})).toEqual({
```
Space here please

benfluleck commented on __tests__/utils/validation.test.js:
```diff
@@ -0,0 +1,116 @@
+      state,
+      mockChangeValue("firstName", "")
+    );
+    expect(formErrors.firstName).toEqual("First name is too short");
```
Space here please

benfluleck commented on src/actions/auth/socialAuth.js:
```diff
@@ -0,0 +1,20 @@
+  return request(`${url}/${socialToken}`, "GET")
+    .then(response => {
+      localStorage.setItem("token", response.token);
+      dispatch({
```
Please put a space here

benfluleck commented on src/actions/auth/verifyUser.js:
```diff
@@ -0,0 +1,21 @@
+  return request(`${API_URL}/auth/verification/${token}`, "GET")
+    .then(response => {
+      localStorage.setItem("token", response.token);
+      dispatch({
```
Please put a space here please

benfluleck commented on src/components/auth/SignUp.jsx:
```diff
@@ -0,0 +1,57 @@
+  handleChange: PropTypes.func,
+  handleSubmit: PropTypes.func,
+  errors: PropTypes.object,
+  authData: PropTypes.object,
```
This shape is not correct but I will just leave you

benfluleck commented:
Also you have a conflict

benfluleck commented on __tests__/components/BaseRoute.test.js:
```diff
@@ -0,0 +1,48 @@
+import ProfilePage from "<pages>/ProfilePage";
+import SocialRedirect from "<auth>/SocialAuth";
+import VerifyAccount from "<auth>/VerifyAccount";
+import store from "<src>/store";
```
Are you testing our actual store? You should use a mock store.
